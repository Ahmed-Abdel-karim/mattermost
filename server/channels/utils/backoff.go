// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package utils

import (
	"time"
)

var backoffTimeouts = []time.Duration{50 * time.Millisecond, 100 * time.Millisecond, 200 * time.Millisecond, 200 * time.Millisecond, 400 * time.Millisecond, 400 * time.Millisecond}

var longBackoffTimeouts = []time.Duration{200 * time.Millisecond, 500 * time.Millisecond, 1 * time.Second, 2 * time.Second, 5 * time.Second, 10 * time.Second}

// ProgressiveRetry executes a BackoffOperation and waits an increasing time before retrying the operation.
func ProgressiveRetry(operation func() error) error {
	//var err error
	//
	//for attempts := 0; attempts < len(backoffTimeouts); attempts++ {
	//	err = operation()
	//	if err == nil {
	//		return nil
	//	}
	//
	//	time.Sleep(backoffTimeouts[attempts])
	//}
	//
	//return err

	return progressiveRetry(operation, backoffTimeouts)
}

func LongProgressiveRetry(operation func() error) error {
	//var err error
	//
	//for attempts := 0; attempts < len(backoffTimeouts); attempts++ {
	//	err = operation()
	//	if err == nil {
	//		return nil
	//	}
	//
	//	time.Sleep(backoffTimeouts[attempts])
	//}
	//
	//return err

	return progressiveRetry(operation, longBackoffTimeouts)
}

func progressiveRetry(operation func() error, timeouts []time.Duration) error {
	var err error

	for attempts := 0; attempts < len(timeouts); attempts++ {
		err = operation()
		if err == nil {
			return nil
		}

		time.Sleep(timeouts[attempts])
	}

	return err
}
