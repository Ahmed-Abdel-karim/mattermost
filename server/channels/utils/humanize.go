// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package utils

import "strings"

func JoinList(items []string) string {
	return strings.Join(items[:len(items)-1], ", ") + " and " + items[len(items)-1]
}
