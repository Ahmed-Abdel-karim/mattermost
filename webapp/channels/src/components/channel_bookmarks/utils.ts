// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {useMemo} from 'react';
import {useSelector} from 'react-redux';

import type {ChannelBookmarksState} from '@mattermost/types/channel_bookmarks';
import type {GlobalState} from '@mattermost/types/store';

import {getFeatureFlagValue, getLicense} from 'mattermost-redux/selectors/entities/general';

import {LicenseSkus} from 'utils/constants';

export const useIsChannelBookmarksEnabled = () => {
    return useSelector(getIsChannelBookmarksEnabled);
};

export const getIsChannelBookmarksEnabled = (state: GlobalState) => {
    const isEnabled = getFeatureFlagValue(state, 'ChannelBookmarks') === 'true';

    if (!isEnabled) {
        return false;
    }

    const license = getLicense(state);

    const isLicensed = license?.IsLicensed === 'true';

    // Channel Bookmarks is available for Professional & Enterprise, and is backward compatible with E20 & E10
    return (
        isLicensed &&
        (
            license.SkuShortName === LicenseSkus.Professional ||
            license.SkuShortName === LicenseSkus.Enterprise ||
            license.SkuShortName === LicenseSkus.E20 ||
            license.SkuShortName === LicenseSkus.E10
        )
    );
};

const EMPTY_BOOKMARKS = {};

export const getChannelBookmarks = (state: GlobalState, channelId: string): ChannelBookmarksState['byChannelId'][string] => {
    const bookmarks = state.entities.channelBookmarks.byChannelId[channelId];

    if (!bookmarks) {
        return EMPTY_BOOKMARKS;
    }

    return bookmarks;
};

export const getChannelBookmark = (state: GlobalState, channelId: string, bookmarkId: string) => {
    return getChannelBookmarks(state, channelId)?.[bookmarkId];
};

export const useChannelBookmarks = (channelId: string) => {
    const bookmarks = useSelector((state: GlobalState) => getChannelBookmarks(state, channelId));

    const order = useMemo(() => {
        return Object.keys(bookmarks).sort((a, b) => bookmarks[a].sort_order - bookmarks[b].sort_order);
    }, [bookmarks]);

    return {
        bookmarks,
        order,
    };
};
