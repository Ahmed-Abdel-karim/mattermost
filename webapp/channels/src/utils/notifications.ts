// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {DispatchFunc} from 'mattermost-redux/types/actions';

import {requestNotificationPermission} from 'actions/notification_actions';

import icon50 from 'images/icon50x50.png';
import iconWS from 'images/icon_WS.png';
import Constants from 'utils/constants';
import * as UserAgent from 'utils/user_agent';

// showNotification displays a platform notification with the configured parameters.
//
// If successful in showing a notification, it resolves with a callback to manually close the
// notification. If no error occurred but the user did not grant permission to show notifications, it
// resolves with a no-op callback. Notifications that do not require interaction will be closed automatically after
// the Constants.DEFAULT_NOTIFICATION_DURATION. Not all platforms support all features, and may
// choose different semantics for the notifications.

export interface ShowNotificationParams {
    title: string;
    body: string;
    requireInteraction: boolean;
    silent: boolean;
    onClick?: (this: Notification, e: Event) => any | null;
}

export function showNotification(
    {
        title,
        body,
        requireInteraction,
        silent,
        onClick,
    }: ShowNotificationParams,
) {
    return async (dispatch: DispatchFunc) => {
        let icon = icon50;
        if (UserAgent.isEdge()) {
            icon = iconWS;
        }

        let permission = Notification.permission;
        console.log('NOTIFCIATION', 'initial permission', permission);
        if (Notification.permission === 'default' && window.isActive) {
            console.log('NOTIFCIATION', 'prompting for permission');
            permission = await Notification.requestPermission();
            console.log('NOTIFCIATION', 'request returned', permission);
        }

        if (permission === 'denied') {
            // User didn't allow notifications
            console.log('NOTIFCIATION', 'permissions denied');
            return () => {};
        } else if (permission === 'default') {
            console.log('NOTIFCIATION', 'showing notification bar');

            // Firefox and Edge doesn't let us prompt to allow notifications except in response to user interaction,
            // so ask for permission the next time they click into the app
            dispatch(requestNotificationPermission());

            return () => {};
        }

        console.log('NOTIFCIATION', 'sending notification');

        const notification = new Notification(title, {
            body,
            tag: body,
            icon,
            requireInteraction,
            silent,
        });

        if (onClick) {
            notification.onclick = onClick;
        }

        notification.onerror = () => {
            throw new Error('Notification failed to show.');
        };

        // Mac desktop app notification dismissal is handled by the OS
        if (!requireInteraction && !UserAgent.isMacApp()) {
            setTimeout(() => {
                notification.close();
            }, Constants.DEFAULT_NOTIFICATION_DURATION);
        }

        return () => {
            notification.close();
        };
    };
}
