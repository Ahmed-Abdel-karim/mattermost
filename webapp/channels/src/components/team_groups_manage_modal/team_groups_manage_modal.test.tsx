// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import type {ComponentProps} from 'react';

import {shallowWithIntl, mountWithIntl} from 'tests/helpers/intl-test-helper';
import {TestHelper} from 'utils/test_helper';

import TeamGroupsManageModal from './team_groups_manage_modal';

describe('components/TeamGroupsManageModal', () => {
    const getGroupsAssociatedToTeam = jest.fn().mockResolvedValue({data: true});
    const closeModal = jest.fn().mockReturnValue({data: true});
    const openModal = jest.fn().mockReturnValue({data: true});
    const unlinkGroupSyncable = jest.fn().mockReturnValue({data: true});
    const patchGroupSyncable = jest.fn().mockReturnValue({data: true});
    const getMyTeamMembers = jest.fn().mockReturnValue({data: true});

    const baseActions = {
        getGroupsAssociatedToTeam,
        closeModal,
        openModal,
        unlinkGroupSyncable,
        patchGroupSyncable,
        getMyTeamMembers,
    };

    const baseProps: ComponentProps<typeof TeamGroupsManageModal> = {
        team: TestHelper.getTeamMock({id: 'team_id'}),
        actions: baseActions,
    };

    test('should match snapshot when groups list is empty', () => {
        const wrapper = shallowWithIntl(<TeamGroupsManageModal {...baseProps}/>);

        expect(wrapper.state('showConfirmModal')).toEqual(false);
        expect(wrapper.state('item')).toEqual({member_count: 0});
        expect(wrapper.state('listModal')).toEqual(undefined);

        expect(wrapper).toMatchSnapshot();
    });

    test('should match snapshot when groups list contains two groups', () => {
        const props = {
            ...baseProps,
            actions: {
                ...baseActions,
                getGroupsAssociatedToTeam: jest.fn().mockResolvedValue(
                    {data: {
                        groups: [
                            TestHelper.getGroupMock({display_name: 'group1'}),
                            TestHelper.getGroupMock({display_name: 'group2'}),
                        ],
                        totalGroupCount: 2,
                    }},
                ),
            },
        };
        const wrapper = shallowWithIntl(<TeamGroupsManageModal {...props}/>);

        expect(wrapper.state('showConfirmModal')).toEqual(false);
        expect(wrapper.state('item')).toEqual({member_count: 0});
        expect(wrapper.state('listModal')).toEqual(undefined);

        expect(wrapper).toMatchSnapshot();
    });

    test('should render using given component', () => {
        const props = {
            ...baseProps,
            actions: {
                ...baseActions,
                getGroupsAssociatedToTeam: jest.fn().mockResolvedValue(
                    {data: {
                        groups: [
                            TestHelper.getGroupMock({display_name: 'group1', id: 'group_id_1'}),
                            TestHelper.getGroupMock({display_name: 'group2', id: 'group_id_2'}),
                        ],
                        totalGroupCount: 2,
                    }},
                ),
            },
        };

        const wrapper = mountWithIntl(
            <TeamGroupsManageModal
                {...props}
            />,
        );

        expect(wrapper.find('.more-modal__name').exists()).toBe(true);
        expect(wrapper.find('.more-modal__name').contains('grop1')).toBe(true);
    });
});
