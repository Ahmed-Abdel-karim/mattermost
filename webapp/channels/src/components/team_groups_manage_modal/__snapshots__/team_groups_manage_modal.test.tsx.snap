// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/TeamGroupsManageModal should match snapshot when groups list contains two groups 1`] = `
<Fragment>
  <ListModal
    loadItems={[Function]}
    onHide={[Function]}
    renderRow={[Function]}
    searchPlaceholderText="Search groups"
    show={true}
    titleBarButtonOnClick={[Function]}
    titleBarButtonText="Add Groups"
    titleText="name Groups"
  />
  <ConfirmModal
    confirmButtonClass="btn btn-primary"
    confirmButtonText="Yes, Remove Group and Members"
    message="0 members associated to this group will be removed from the team. Are you sure you wish to remove this group and 0 members?"
    modalClass=""
    onCancel={[Function]}
    onConfirm={[Function]}
    show={false}
    title="Remove Group and 0 Members"
  />
</Fragment>
`;

exports[`components/TeamGroupsManageModal should match snapshot when groups list is empty 1`] = `
<Fragment>
  <ListModal
    loadItems={[Function]}
    onHide={[Function]}
    renderRow={[Function]}
    searchPlaceholderText="Search groups"
    show={true}
    titleBarButtonOnClick={[Function]}
    titleBarButtonText="Add Groups"
    titleText="name Groups"
  />
  <ConfirmModal
    confirmButtonClass="btn btn-primary"
    confirmButtonText="Yes, Remove Group and Members"
    message="0 members associated to this group will be removed from the team. Are you sure you wish to remove this group and 0 members?"
    modalClass=""
    onCancel={[Function]}
    onConfirm={[Function]}
    show={false}
    title="Remove Group and 0 Members"
  />
</Fragment>
`;
